openebs/pkg/volume/provision_v1alpha1.go

```diff
@@ -0,0 +1,201 @@
+
+	nodeName := os.Getenv("NODE_NAME")
+	if nodeName == "" {
+		glog.Errorf("ENV variable 'NODE_NAME' is not set")
+	}
+	var openebsObj mApiv1.OpenEBSVolume
+
+	//Get maya-apiserver IP address from cluster
+	addr, err := openebsObj.GetMayaClusterIP(client)
+
+	if err != nil {
+		glog.Errorf("Error getting maya-apiserver IP Address: %v", err)
+		return nil
+	}
+	mayaServiceURI := "http://" + addr + ":5656"
+
+	//Set maya-apiserver IP address along with default port
+	os.Setenv("MAPI_ADDR", mayaServiceURI)
+
+	return &openEBSProvisionerV1alpha1{
+		mapiURI:  mayaServiceURI,
+		identity: nodeName,
+	}
+}
+
+var _ controller.Provisioner = &openEBSProvisionerV1alpha1{}
+
+// Provision creates a storage asset and returns a PV object representing it.
+func (p *openEBSProvisionerV1alpha1) Provision(options controller.VolumeOptions) (*v1.PersistentVolume, error) {
+
+	//Issue a request to Maya API Server to create a volume
+	var volume mayav1.Volume
+	var openebsCASVol mApiv1.OpenEBSVolumeV1Alpha1
+	casVolume := v1alpha1.CASVolume{}
+
+	volSize := options.PVC.Spec.Resources.Requests[v1.ResourceName(v1.ResourceStorage)]
+	casVolume.Spec.Capacity = volSize.String()
+
+	className := GetStorageClassName(options)
+
+	// creating a map b/c have to initialize the map using the make function before
+	// adding any elements to avoid nil map assignment error
+	mapLabels := make(map[string]string)
+
+	if className == nil {
+		glog.Errorf("Volume has no storage class specified")
+	} else {
+		mapLabels[string(v1alpha1.StorageClassKey)] = *className
+		glog.Infof("Using Storageclass: %s", *className)
+		casVolume.Labels = mapLabels
+	}
+	casVolume.Labels[string(v1alpha1.NamespaceKey)] = options.PVC.Namespace
+	casVolume.Namespace = options.PVC.Namespace
+	casVolume.Labels[string(v1alpha1.PersistentVolumeClaimKey)] = options.PVC.ObjectMeta.Name
+	casVolume.Name = options.PVName
+
+	//Pass through labels from PVC to maya-apiserver
+	//volumeSpec.Metadata.Labels.Application = options.PVC.ObjectMeta.GetLabels()[mayav1.PVCLabelsApplication]
+	//volumeSpec.Metadata.Labels.ReplicaTopoKeyDomain = options.PVC.ObjectMeta.GetLabels()[mayav1.PVCLabelsReplicaTopKeyDomain]
+	//volumeSpec.Metadata.Labels.ReplicaTopoKeyType = options.PVC.ObjectMeta.GetLabels()[mayav1.PVCLabelsReplicaTopKeyType]
+
+	_, err := openebsCASVol.CreateVolume(casVolume)
+	if err != nil {
+		glog.Errorf("Error creating volume: %v", err)
+		return nil, err
+	}
+
+	err = openebsCASVol.ReadVolume(options.PVName, options.PVC.Namespace, &volume)
+	if err != nil {
+		glog.Errorf("Error getting volume details: %v", err)
+		return nil, err
+	}
+
+	// Use annotations to specify the context using which the PV was created.
+	volAnnotations := make(map[string]string)
+	volAnnotations["openEBSProvisionerIdentity"] = p.identity
+
+	var iqn, targetPortal string
+
+	for key, value := range volume.Metadata.Annotations.(map[string]interface{}) {
```

Comment:
we will not use annotations. Let us create properties in CASVolume.
This needs to be done at maya api server as well as here.